### ISSUE TITLE
Classmethod returning `Self` makes Pylance raise `reportReturnType`

### ISSUE THREAD

**Issue by simensol** [microsoft/pyright#7251]
My Django app is typed with [sbdchd/django-types](https://github.com/sbdchd/django-types), the same types which are shipped with Pylance.

When my custom managers went from untyped:

```py
objects  = GroupManager()
```

to typed:

```py
objects: ClassVar[GroupManager] = GroupManager()
```

classmethods returning `Self`, e.g.:

```py
@classmethod
def create(
    cls,
    name: str
) -> Self:
    return cls.objects.create(
        name=name,
    )
```

started to raise the following linting error:

```bash
Expression of type "Group" cannot be assigned to return type "Self@Group"
  Type "Group" cannot be assigned to type "Self@Group" Pylance(reportReturnType)
```

I fixed this by returning `"Group"` instead:

```py
@classmethod
def create(
    cls,
    name: str
) -> "Group":
    return cls.objects.create(
        name=name,
    )
```

However, according to [PEP 673](https://peps.python.org/pep-0673/#use-in-classmethod-signatures), a classmethod should be able to return `Self`.

It happens in both Pylance `v2024.2.1` and when I run Pyright `v1.1.349` from the terminal.

I have reported the issue in https://github.com/sbdchd/django-types/issues/226, but the repro owner suggested that this was a Pyright issue.

**Comment by simensol** [comment]
Sure, here is a full code sample:

```py
from __future__ import annotations

from typing import ClassVar, Self

from django.db import models


class TestManager(models.Manager["TestModel"]): ...


class TestModel(models.Model):
    objects: ClassVar[TestManager] = TestManager()

    @classmethod
    def create(cls, name: str) -> Self:
        return cls.objects.create(
            name=name,
        )
```

which raises

```bash
Expression of type "TestModel" cannot be assigned to return type "Self@TestModel"
  Type "TestModel" cannot be assigned to type "Self@TestModel" Pylance(reportReturnType)
```

<img width="837" alt="backend" src="https://github.com/microsoft/pyright/assets/4491653/c8956852-3441-4790-a589-38743ea1492a">

Replacing `objects: ClassVar[TestManager] = TestManager()` with `objects = TestManager()` removed the type issue (see https://github.com/sbdchd/django-types/issues/194#issuecomment-1917046748), but raises a new one for the the `objects = TestManager()` line:

<img width="887" alt="backend" src="https://github.com/microsoft/pyright/assets/4491653/36eefd01-1f9e-43c2-a438-8116974a5f73">

**Comment by erictraut** [comment]
Thanks for the complete code sample.

Pyright is behaving correctly here from the perspective of the Python type spec, so I don't consider this a bug.

Here's a way you can avoid the type violation:

```python
from __future__ import annotations

from typing import Self, TypeVar
from django.db import models

T = TypeVar("T", bound="TestModel")

class TestManager(models.Manager[T]): ...

class TestModel(models.Model):
    objects = TestManager[Self]()

    @classmethod
    def create(cls, name: str) -> Self:
        return cls.objects.create(name=name)
```

Alternatively:

```python
from __future__ import annotations

from typing import ClassVar
from django.db import models

class TestManager(models.Manager["TestModel"]): ...

class TestModel(models.Model):
    objects: ClassVar[models.BaseManager["TestModel"]] = TestManager()

    @classmethod
    def create(cls, name: str) -> TestModel:
        return cls.objects.create(name=name)
```

I'd prefer the first of these two because it will work better if you ever subclass `TestModel`.

**Comment by simensol** [comment]
For later reference: With the first (preferable) solution, I had to add types to the `objects` variable for typing of manager methods (e.g. `TestModel.objects.get_some_object()`) to work:

```python
class TestModel(models.Model):
    objects: ClassVar[TestManager[Self]] = TestManager()
```
